File: android-app/app/src/main/java/com/bookstore/bookapp/presentation/viewmodel/ChatViewModel.kt
package com.bookstore.bookapp.presentation.viewmodel

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bookstore.bookapp.data.remote.api.Message
import com.bookstore.bookapp.data.remote.socket.ChatSocketManager
import com.bookstore.bookapp.data.remote.socket.SocketConnectionState
import com.bookstore.bookapp.data.remote.socket.TypingEvent
import com.bookstore.bookapp.di.TokenHolder
import com.bookstore.bookapp.domain.model.Transaction
import com.bookstore.bookapp.domain.model.TransactionStatus
import com.bookstore.bookapp.domain.repository.AuthRepository
import com.bookstore.bookapp.domain.repository.TransactionRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ChatState(
    val isLoading: Boolean = true,
    val transaction: Transaction? = null,
    val messages: List<Message> = emptyList(),
    val currentUserId: String? = null,
    val error: String? = null,
    val isSending: Boolean = false,
    val chatAvailable: Boolean = false,
    val isSocketConnected: Boolean = false,
    val otherUserTyping: Boolean = false
) {
    val otherPartyName: String
        get() {
            if (transaction == null || currentUserId == null) return ""
            return if (transaction.borrowerId == currentUserId) 
                transaction.ownerName 
            else 
                transaction.borrowerName
        }

    val otherPartyImageUrl: String?
        get() {
            if (transaction == null || currentUserId == null) return null
            return if (transaction.borrowerId == currentUserId)
                transaction.ownerProfileImageUrl
            else
                transaction.borrowerProfileImageUrl
        }
}

@HiltViewModel
class ChatViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val transactionRepository: TransactionRepository,
    private val authRepository: AuthRepository,
    private val chatSocketManager: ChatSocketManager,
    private val tokenHolder: TokenHolder
) : ViewModel() {

    private val transactionId: String? = savedStateHandle["transactionId"]

    private val _uiState = MutableStateFlow(ChatState())
    val uiState: StateFlow<ChatState> = _uiState.asStateFlow()

    private var pollingJob: Job? = null
    private var usePollingFallback = false

    init {
        loadChat()
        observeSocketConnection()
        observeIncomingMessages()
        observeTypingEvents()
    }

    private fun loadChat() {
        val id = transactionId ?: run {
            _uiState.value = _uiState.value.copy(isLoading = false, error = "Invalid transaction")
            return
        }

        viewModelScope.launch {
            try {
                val user = authRepository.currentUser.first()
                val transaction = transactionRepository.fetchTransactionById(id)
                val chatAvailable = transaction.status == TransactionStatus.APPROVED || 
                                   transaction.status == TransactionStatus.ACTIVE

                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    transaction = transaction,
                    currentUserId = user?.id,
                    chatAvailable = chatAvailable,
                    error = null
                )

                if (chatAvailable) {
                    loadMessages()
                    connectToSocket()
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    error = e.localizedMessage ?: "Failed to load chat"
                )
            }
        }
    }

    private fun connectToSocket() {
        val token = tokenHolder.accessToken
        if (token != null) {
            chatSocketManager.connect(token)
            transactionId?.let { chatSocketManager.joinChat(it) }
        } else {
            startPollingFallback()
        }
    }

    private fun observeSocketConnection() {
        viewModelScope.launch {
            chatSocketManager.connectionState.collect { state ->
                val isConnected = state is SocketConnectionState.Connected
                _uiState.value = _uiState.value.copy(isSocketConnected = isConnected)
                
                when (state) {
                    is SocketConnectionState.Connected -> {
                        stopPollingFallback()
                        transactionId?.let { chatSocketManager.joinChat(it) }
                    }
                    is SocketConnectionState.Disconnected,
                    is SocketConnectionState.Error -> {
                        if (_uiState.value.chatAvailable) {
                            startPollingFallback()
                        }
                    }
                    is SocketConnectionState.Connecting -> {
                    }
                }
            }
        }
    }

    private fun observeIncomingMessages() {
        viewModelScope.launch {
            chatSocketManager.incomingMessages.collect { message ->
                if (message.transactionId == transactionId) {
                    val currentMessages = _uiState.value.messages
                    if (currentMessages.none { it.id == message.id }) {
                        _uiState.value = _uiState.value.copy(
                            messages = currentMessages + message
                        )
                    }
                }
            }
        }
    }

    private fun observeTypingEvents() {
        viewModelScope.launch {
            chatSocketManager.typingEvents.collect { event ->
                if (event.transactionId == transactionId && 
                    event.userId != _uiState.value.currentUserId) {
                    _uiState.value = _uiState.value.copy(otherUserTyping = event.isTyping)
                }
            }
        }
    }

    private fun loadMessages() {
        val id = transactionId ?: return
        viewModelScope.launch {
            try {
                val messages = transactionRepository.getMessages(id)
                _uiState.value = _uiState.value.copy(messages = messages)
            } catch (e: Exception) {
            }
        }
    }

    private fun startPollingFallback() {
        if (pollingJob?.isActive == true) return
        usePollingFallback = true
        
        pollingJob = viewModelScope.launch {
            while (isActive && usePollingFallback) {
                delay(5000)
                if (_uiState.value.chatAvailable && !_uiState.value.isLoading) {
                    loadMessages()
                }
            }
        }
    }

    private fun stopPollingFallback() {
        usePollingFallback = false
        pollingJob?.cancel()
        pollingJob = null
    }

    fun sendMessage(content: String) {
        val id = transactionId ?: return
        if (content.isBlank()) return

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isSending = true)
            try {
                val newMessage = transactionRepository.sendMessage(id, content)
                val currentMessages = _uiState.value.messages
                if (currentMessages.none { it.id == newMessage.id }) {
                    _uiState.value = _uiState.value.copy(
                        isSending = false,
                        messages = currentMessages + newMessage
                    )
                } else {
                    _uiState.value = _uiState.value.copy(isSending = false)
                }
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isSending = false,
                    error = e.localizedMessage ?: "Failed to send message"
                )
            }
        }
    }

    fun onTypingStateChanged(isTyping: Boolean) {
        transactionId?.let { id ->
            chatSocketManager.sendTypingStatus(id, isTyping)
        }
    }

    fun refresh() {
        loadMessages()
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    override fun onCleared() {
        super.onCleared()
        transactionId?.let { chatSocketManager.leaveChat(it) }
        stopPollingFallback()
    }
}

File: bookApp/bookApp/bookApp/ViewModels/ChatViewModel.swift
import Foundation
import Combine

@MainActor
class ChatViewModel: ObservableObject {
    @Published private(set) var messages: [Message] = []
    @Published var isLoading = false
    @Published var error: String?
    @Published var newMessageText: String = ""
    @Published var transaction: Transaction?
    @Published var otherPartyName: String = "Chat"
    @Published var otherPartyImageUrl: String?
    @Published var isSending = false
    @Published var chatAvailable = true
    @Published var isSocketConnected = false
    @Published var otherUserTyping = false

    let transactionId: String
    private let store = AppDataStore.shared
    private let messageService = MessageService.shared
    private let transactionService = TransactionService()
    private let socketManager = ChatSocketManager.shared
    private var usePollingFallback = false
    private var pollTimer: Timer?
    private var cancellables = Set<AnyCancellable>()
    private var currentUserId: String?
    
    init(transactionId: String) {
        self.transactionId = transactionId
        setupMessageObservation()
        setupSocketObservation()
    }
    
    private func setupMessageObservation() {
        store.$messagesByTransaction
            .map { $0[self.transactionId] ?? [] }
            .receive(on: DispatchQueue.main)
            .assign(to: \.messages, on: self)
            .store(in: &cancellables)
    }
    
    private func setupSocketObservation() {
        socketManager.$connectionState
            .receive(on: DispatchQueue.main)
            .sink { [weak self] state in
                guard let self = self else { return }
                self.isSocketConnected = (state == .connected)
                
                switch state {
                case .connected:
                    self.stopPollingFallback()
                    self.socketManager.joinChat(transactionId: self.transactionId)
                case .disconnected, .error:
                    if self.chatAvailable {
                        self.startPollingFallback()
                    }
                case .connecting:
                    break
                }
            }
            .store(in: &cancellables)
        
        socketManager.incomingMessages
            .filter { [weak self] message in
                message.transactionId == self?.transactionId
            }
            .receive(on: DispatchQueue.main)
            .sink { [weak self] message in
                guard let self = self else { return }
                if !self.messages.contains(where: { $0.id == message.id }) {
                    var updatedMessages = self.messages
                    updatedMessages.append(message)
                    self.store.storeMessages(updatedMessages, transactionId: self.transactionId)
                }
            }
            .store(in: &cancellables)
        
        socketManager.typingEvents
            .filter { [weak self] event in
                event.transactionId == self?.transactionId &&
                event.userId != self?.currentUserId
            }
            .receive(on: DispatchQueue.main)
            .sink { [weak self] event in
                self?.otherUserTyping = event.isTyping
            }
            .store(in: &cancellables)
    }

    func fetchInitialData(currentUserId: String) async {
        self.currentUserId = currentUserId
        isLoading = true
        error = nil
        
        do {
            if let cached = store.borrowedTransactions.first(where: { $0.id == transactionId }) {
                self.transaction = cached
            } else if let cached = store.ownerTransactions.first(where: { $0.id == transactionId }) {
                self.transaction = cached
            }
            
            if let t = self.transaction {
                updateOtherPartyInfo(t, currentUserId: currentUserId)
            }

            let fetchedTransaction = try await transactionService.fetchTransactionById(id: transactionId)
            self.transaction = fetchedTransaction
            
            updateOtherPartyInfo(fetchedTransaction, currentUserId: currentUserId)
            
            let status = fetchedTransaction.status
            self.chatAvailable = (status == .pending || status == .approved || status == .active)

            if chatAvailable {
                _ = await messageService.fetchCachedMessages(transactionId: transactionId)
                _ = try await messageService.fetchMessages(transactionId: transactionId)
                await markAsRead()
                
                connectToSocket()
            }
        } catch {
            self.error = error.localizedDescription
        }
        
        isLoading = false
    }
    
    private func connectToSocket() {
        if let token = KeychainManager.shared.getAccessToken() {
            socketManager.connect(token: token)
            socketManager.joinChat(transactionId: transactionId)
        } else {
            startPollingFallback()
        }
    }

    func fetchMessages() async {
        guard chatAvailable else { return }
        do {
            _ = try await messageService.fetchMessages(transactionId: transactionId)
            await markAsRead()
        } catch {
            self.error = error.localizedDescription
        }
    }
    
    func sendMessage() async {
        let text = newMessageText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !text.isEmpty, chatAvailable, !isSending else { return }
        
        isSending = true
        socketManager.sendTypingStatus(transactionId: transactionId, isTyping: false)
        
        do {
            let sentMessage = try await messageService.sendMessage(transactionId: transactionId, content: text)
            
            if !messages.contains(where: { $0.id == sentMessage.id }) {
                var updatedMessages = self.messages
                updatedMessages.append(sentMessage)
                store.storeMessages(updatedMessages, transactionId: transactionId)
            }
            
            self.newMessageText = ""
        } catch {
            self.error = error.localizedDescription
        }
        isSending = false
    }
    
    func markAsRead() async {
        do {
            try await messageService.markAsRead(transactionId: transactionId)
            socketManager.sendMessageRead(transactionId: transactionId)
        } catch {
            print("Failed to mark messages as read: \(error.localizedDescription)")
        }
    }
    
    func onTypingStateChanged(_ isTyping: Bool) {
        socketManager.sendTypingStatus(transactionId: transactionId, isTyping: isTyping)
    }
    
    private func startPollingFallback() {
        guard pollTimer == nil, chatAvailable else { return }
        usePollingFallback = true
        
        pollTimer = Timer.scheduledTimer(withTimeInterval: 5.0, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in
                guard let self = self, self.usePollingFallback else { return }
                await self.fetchMessages()
            }
        }
    }
    
    private func stopPollingFallback() {
        usePollingFallback = false
        pollTimer?.invalidate()
        pollTimer = nil
    }
    
    func startPolling() {
        startPollingFallback()
    }
    
    func stopPolling() {
        socketManager.leaveChat(transactionId: transactionId)
        stopPollingFallback()
    }

    private func updateOtherPartyInfo(_ t: Transaction, currentUserId: String) {
        if t.ownerId == currentUserId {
            self.otherPartyName = t.borrowerName
            self.otherPartyImageUrl = t.borrowerProfileImageUrl
        } else {
            self.otherPartyName = t.ownerName
            self.otherPartyImageUrl = t.ownerProfileImageUrl
        }
    }
}
